### DISCUSSION TITLE
Testing a package's public API

### DISCUSSION THREAD

**Discussion by altano** [vitest-dev/vitest#5368] [Feedback]
The default way of using Vitest is to just directly import the file you are testing with a relative path, which obviously works.

There are some people (e.g. https://fosstodon.org/@rauschma/112066537954393244, see comments) who would like to test the public API of their package. In Node.js this is possible by just importing whatever you want to test by referencing your package name and your package.json "exports". It would look like this:

```package.json
{
  "name": "my-awesome-fn",
  "type": "module",
  "exports": "./dist/awesomeFn.js"
}
```

```my-test.ts
import awesomeFn from "my-awesome-fn";
```

@Ifiokjr asked about this feature in https://github.com/vitest-dev/vitest/issues/1394 but @antfu closed that as "possible with `resolve.alias`". The problem with using `resolve.alias` is that the main use case of self-referencing your package is to test the package through the public API, and if you're manually replicating the behavior of package.json's `exports` field in `resolve.alias` you're not really testing the public API anymore.

Please reconsider implementing the specced Node.js module resolution in Vitest that allows self-referencing the package via package.json exports.

Thank you!

**Comment by hi-ogawa** [comment]
>  the main use case of self-referencing your package is to test the package through the public API, and if you're manually replicating the behavior of package.json's exports field in resolve.alias you're not really testing the public API anymore.

I understand this perspective, but Vite and Vitest's import resolutions and transforms are too forgiving compared to actual NodeJS runtime in many aspects, so if you need to test pure NodeJS use case, then Vitest is not technically a good fit (unless Vitest provides such stricter mode https://github.com/vitest-dev/vitest/issues/3849).

Also, maybe this self-reference issue is simply a Vite bug and it looks like this is going to be fixed soon:
- https://github.com/vitejs/vite/issues/9731

**Reply by altano** [reply]
Thank you for replying. Yes, I understand your perspective as well: it's probably folly to rely on Vitest perfectly replicating the Node.js module resolution algorithm. I just hope this is one hole that can get addressed, given the specific and useful use case presented.

And thanks for that link, I missed it! I'm curious why you say it looks like it is going to be fixed soon, because the discussion in  vitejs/vite#9731 doesn't seem to mention any potential fixes being on the way.

**Reply by hi-ogawa** [reply]
The issue is very old one, but there's a PR coming https://github.com/vitejs/vite/pull/16068 and it's already approved by the team.

**Reply by altano** [reply]
@hi-ogawa since this was fixed and released in vite 5.2.0 (https://github.com/vitejs/vite/blob/v5.2.0/packages/vite/CHANGELOG.md#520-beta0-2024-03-12) can I expect this to work in the next release of vitest, whenever that happens to be?

Or is there something on the Vitest side that needs doing as well?

Thanks again!

**Reply by hi-ogawa** [reply]
Vitest doesn't pin Vite's version, so if you explicitly install Vite 5.2 for your project, then Vitest will use that one. Or you may be able to just upgrade transitive dependencies, for example, by `pnpm dedupe` https://pnpm.io/cli/dedupe

https://github.com/vitest-dev/vitest/blob/c62b3a212073c8a81a617865bc3ba0c14a4fb632/packages/vitest/package.json#L163

**Comment by altano** [comment]
This has shipped in Vite 5.2.0 and can be used in Vitest today, so I'm closing this discussion.

Thanks so much @hi-ogawa for all your help!
